include/util/json_renderer.hpp

```diff
@@ -27,18 +27,32 @@ namespace
 constexpr int MAX_FLOAT_STRING_LENGTH = 256;
 }
 
-struct Renderer
+template <typename Out> struct Renderer
 {
-    explicit Renderer(std::ostream &_out) : out(_out) {}
+    explicit Renderer(Out &_out) : out(_out) {}
 
-    void operator()(const String &string) const
+    void operator()(const String &string)
     {
-        out << "\"";
-        out << escape_JSON(string.value);
-        out << "\"";
+        write('"');
+        // here we assume that vast majority of strings don't need to be escaped,
+        // so we check it first and escape only if needed
+        auto size = SizeOfEscapedJSONString(string.value);
+        if (size == string.value.size())
+        {
+            write(string.value);
+        }
+        else
+        {
+            std::string escaped;
+            escaped.reserve(size);
+            EscapeJSONString(string.value, escaped);
+
+            write(escaped);
+        }
+        write('"');
     }
 
-    void operator()(const Number &number) const
+    void operator()(const Number &number)
```

Comment:
We spend quite a lot on number formatting: I've tried to optimize it using `fmt`(which promises to have faster formatting than current algorithm), but I couldn't notice the difference.

Reply:
Yeah - I optimized this part a long time ago - if you dig into this function, you'll notice we're _already_ using a highly optimized floating point -> string rendering implementation, rather than just doing `std::to_string(number)`.

Turning floats into strings is fairly expensive, which is why it still shows up in a profiler run.  There are a lot of papers about number->string rendering performance, you can _possibly_ find a better implementation than the one already here, but the gains over what we have will likely be small.

Reply:
Yes, but as I can see we currently use `milo` algorithm(did you modify it in some way btw?) and `fmtlib` author [says their algorithm is even faster](https://github.com/fmtlib/dtoa-benchmark).
<img width="673" alt="Screenshot 2022-09-30 at 18 33 04" src="https://user-images.githubusercontent.com/266271/193316350-9d0a6479-3e2b-4392-a733-91b76251a213.png">

It seemed to be low-hanging fruit to use it here(we already depend on fmtlib anyway), but 100% agree - no reason to waste time on it since possible improvement is marginal.

Reply:
It seems we have UB in current implementation 🤔 
<img width="1960" alt="Screenshot 2022-09-30 at 18 36 56" src="https://user-images.githubusercontent.com/266271/193316976-2e757dda-7ee0-41f6-afc3-b5ffca5f60b6.png">

https://github.com/Project-OSRM/osrm-backend/actions/runs/3160140786/jobs/5144255047

It seems we had no UBSAN alerts here before, because for `std::vector` `ArrayRenderer` was used, which used `std::stringstream` for float numbers rendering.

Reply:
Checked on separate PR that `Renderer` has this UB problem https://github.com/Project-OSRM/osrm-backend/pull/6381
So it seems to be another reason to switch to `fmt`...

Reply:
# 5.

With `fmt` is feels as it is slightly faster(but hard to notice the difference without statistical tests - I am judging just by presence of < 10ms measurements) and no UBs at the same time :)

<img width="1226" alt="Screenshot 2022-10-01 at 21 35 10" src="https://user-images.githubusercontent.com/266271/193425482-82fb3330-bc54-41b2-af35-73d125d632bc.png">